src/QueryAccess/ConditionGroup.php

```diff
@@ -0,0 +1,227 @@
+use Drupal\Core\Cache\RefinableCacheableDependencyTrait;
+
+/**
+ * Represents a group of access conditions.
+ *
+ * Used by query access handlers for filtering lists of entities based on
+ * granted permissions.
+ *
+ * Examples:
+ * @code
+ *   // Filter by node type and uid.
+ *   $condition_group = new ConditionGroup();
+ *   $condition_group->addCondition('type', ['article', 'page']);
+ *   $condition_group->addCondition('uid', '1');
+ *
+ *   // Filter by node type or status.
+ *   $condition_group = new ConditionGroup('OR');
+ *   $condition_group->addCondition('type', ['article', 'page']);
+ *   $condition_group->addCondition('status', '1', '<>');
+ *
+ *   // Nested condition groups: node type AND (uid OR status).
+ *   $condition_group = new ConditionGroup();
+ *   $condition_group->addCondition('type', ['article', 'page']);
+ *   $condition_group->addCondition((new ConditionGroup('OR'))
+ *     ->addCondition('uid', 1)
+ *     ->addCondition('status', '1')
+ *   );
+ * @endcode
+ */
+final class ConditionGroup implements \Countable, RefinableCacheableDependencyInterface {
+
+  use RefinableCacheableDependencyTrait;
+
+  /**
+   * The conjunction.
+   *
+   * @var string
+   */
+  protected $conjunction;
+
+  /**
+   * The conditions.
+   *
+   * @var \Drupal\entity\QueryAccess\Condition[]|\Drupal\entity\QueryAccess\ConditionGroup[]
+   */
+  protected $conditions = [];
+
+  /**
+   * Whether the condition group is always FALSE.
+   *
+   * @var bool
+   */
+  protected $alwaysFalse = FALSE;
+
+  /**
+   * Constructs a new ConditionGroup object.
+   *
+   * @param string $conjunction
+   *   The conjunction.
+   */
+  public function __construct($conjunction = 'AND') {
+    $this->conjunction = $conjunction;
+  }
+
+  /**
+   * Gets the conjunction.
+   *
+   * @return string
+   *   The conjunction. Possible values: AND, OR.
+   */
+  public function getConjunction() {
+    return $this->conjunction;
+  }
+
+  /**
+   * Gets all conditions and nested condition groups.
+   *
+   * @return \Drupal\entity\QueryAccess\Condition[]|\Drupal\entity\QueryAccess\ConditionGroup[]
+   *   The conditions, where each one is either a Condition or a nested
+   *   ConditionGroup. Returned by reference, to allow callers to replace
+   *   or remove conditions.
+   */
+  public function &getConditions() {
+    return $this->conditions;
+  }
+
+  /**
+   * Adds a condition.
+   *
+   * @param string|\Drupal\entity\QueryAccess\ConditionGroup $field
+   *   Either a condition group (for nested AND/OR conditions), or a
+   *   field name with an optional column name. E.g: 'uid', 'address.locality'.
+   * @param mixed $value
+   *   The value.
+   * @param string $operator
+   *   The operator.
+   *   Possible values: =, <>, <, <=, >, >=, BETWEEN, NOT BETWEEN,
+   *                   IN, NOT IN, IS NULL, IS NOT NULL.
+   *
+   * @return $this
+   */
+  public function addCondition($field, $value = NULL, $operator = NULL) {
```

Comment:
Because I could not figure out a decent DX without it. Gabe has a proposed PR linked above that tried to remove mutability, but introduced a DX regression. Would probably be better to discuss on Slack.

Reply:
I tried, and failed too.

The similarity between interactions with `ConditionGroup` and `AccessResult` seem reasonable. I think that makes sense. So I'm fine with keeping this as-is.

Reply:
Consider this resolved.